challenge_serving_drinks/src/challenge_serving_drinks/sd_states.py

```diff
@@ -0,0 +1,539 @@
+        :raises: (TimeoutException) if robot didn't hear anything
+        """
+        attempt = 0
+        while not rospy.is_shutdown():
+            attempt += 1
+            try:
+                speech_result = self._robot.hmi.query(description="Which is the unavailable beverage?",
+                                                      grammar=self._drinks_grammar,
+                                                      target=self._drinks_target)
+                return speech_result
+            except TimeoutException:
+                if attempt >= max_tries:
+                    raise  # Re-raise
+
+            # We only end up here if the exception was caught
+            self._robot.speech.speak(random.choice([
+                "I'm sorry, can you repeat",
+                "Please repeat the drink, I didn't hear you",
+                "I didn't get the drink, can you repeat it",
+                "Please speak up, as I didn't hear what you just said",
+            ]))
+
+    def _confirm(self):
+        """
+        Queries the HMI server for confirmation
+        :return: (bool) whether the understood beverage was correct
+        """
+        try:
+            speech_result = self._robot.hmi.query(description="Is this correct?", grammar="T[True] -> yes;"
+                                                                                          "T[False] -> no", target="T")
+        except TimeoutException:
+            return False
+
+        return speech_result.semantics
+
+
+class DetectWaving(smach.State):
+    """ Waits for waving person.
+
+    This is based on the 'WaitForCustomer' class of the the restaurant challenge. Might
+    be nice to merge these two.
+    """
+    def __init__(self, robot, caller_id):
+        """ Constructor
+
+        :param robot: robot object
+        """
+        smach.State.__init__(self, outcomes=['succeeded', 'aborted'])
+        self._robot = robot
+        self._caller_id = caller_id
+        self._people_sub = rospy.Subscriber(robot.robot_name + '/persons', People, self.people_cb, queue_size=1)
+        self.people_queue = Queue(maxsize=1)
+
+    def people_cb(self, persons):
+        if persons.people:
+            rospy.logdebug('Received %d persons in the people cb', len(persons.people))
+        self.people_queue.put(persons)
+
+    def execute(self, userdata=None):
+
+        self._robot.head.reset()
+        self._robot.head.wait_for_motion_done()
+
+        self._robot.speech.speak("Please wave if you want me to bring you something")
+
+        head_samples = 20
+        look_distance = 3.0
+        look_angles = [0,
+                       0,
+                       0,
+                       10,
+                       -10,
+                       20,
+                       -20]
+        self.clear_queue()
+
+        waving_persons = []
+        i = 0
+        while not rospy.is_shutdown() and not waving_persons:
```

Comment:
It looks like this goes on forever if no waving person is found. Is that really the desired behaviour.

Reply:
The state is not used currently, it's still work in progress. 
In the challenge scenario, the detection is skipped for now but will be updated.

Reply:
Please add some comments about it